batcher.go

```diff
@@ -0,0 +1,151 @@
+package core
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/Hatch1fy/errors"
+)
+
+func newBatcher(core *Core, opts *Opts) *batcher {
+	var b batcher
+	b.core = core
+	b.opts = opts
+	return &b
+}
+
+type batcher struct {
+	mux sync.Mutex
+
+	core *Core
+	opts *Opts
+
+	timer *time.Timer
+	calls []call
+}
+
+func (b *batcher) performCalls(txn *Transaction, cs calls) (failIndex int, err error) {
+	failIndex = -1
+	for i, c := range cs {
+		if err = recoverCall(txn, c.fn); err != nil {
+			failIndex = i
+			return
+		}
+	}
+
+	return
+}
+
+func (b *batcher) clearTimer() {
+	if b.timer == nil {
+		return
+	}
+
+	// Stop timer
+	b.timer.Stop()
+
+	// Clear timer
+	b.timer = nil
+}
+
+// run performs the transactions in the batch and communicates results
+// back to DB.Batch.
+func (b *batcher) run(cs calls) {
+	if len(cs) == 0 {
+		// We have no calls to run, bail out
+		return
+	}
+
+	var failIndex int
+	err := b.core.Transaction(func(txn *Transaction) (err error) {
+		failIndex, err = b.performCalls(txn, cs)
+		return
+	})
+
+	if err == errors.ErrIsClosed {
+		cs.notifyAll(err)
+		return
+	}
+
+	// Check to see if we had no failures in our batch
+	if failIndex == -1 {
+		// We successfully batched our list of calls without error, notify all calls of nil error status
+		cs.notifyAll(nil)
+		return
+	}
+
+	// Create group for successful calls
+	successful := cs[:failIndex]
+
+	// Attempt to retry the successful group before the failing call
+	b.retry(successful, err)
+
+	// Send error down error channel to call who caused issue
+	cs[failIndex].errC <- err
+
+	// Create group for remaining calls
+	remaining := cs[failIndex+1:]
```

Comment:
We should introduce failing calls in our benchmarks to see how this affects performance

Reply:
We can do that later I think